### PR TITLE
feat: rfc: pytorch primitives api, another option

### PR DESCRIPTION
**Pull request by rb-determined-ai**
## Description

Compared to the proposed API of Pytorch Flexible Primitives, I don't see any ways that this API is any more or less capable; that API is definitely not bad.  I just had some ideas of how we support the same features in a way that was hopefully easier for users, and I wanted to post this to see what other people thought.

### CODE REVIEW

File: pytorch_multi_things_2.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        # Users clip gradients in callbacks specified in the scaler.step() operation
+        # (this is because we handle gradient aggregation ourselves)
+        def my_clip_grads_0(...):
+            return ...
+
+        def my_clip_grads_1(...):
+            return ...
+
+        self.scaler.step(self.optimizer0, clip_grads=my_clip_grads_0)
+        self.scaler.step(self.optimizer1, clip_grads=my_clip_grads_1)
```
Comment (katport) [inline pytorch_multi_things_2.py:160]: I like the idea of it being similar to the pytorch-native version. It allows for easier porting, and allows the user the controls they want. I think we could handle the optim.zero_grad ourselves especially since we handle gradient aggregation.

File: pytorch_multi_things_2.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        input, target = batch
+
+        output0 = self.model0(input)
+        output1 = self.model1(input)
+        loss0 = loss_fn(2 * output0 + 3 * output1, target)
+        loss1 = loss_fn(3 * output0 - 5 * output1, target)
+
+        # TODO: name this better, but it is analagous to the scaler.scale() function
+        self.context.loss_tensor(loss0).backward(retain_graph=True)
+        self.context.loss_tensor(loss1).backward(retain_graph=True)
```
Comment (rb-determined-ai) [inline pytorch_multi_things_2.py:207]: typo: this second retain_graph=True is unnecessary

File: pytorch_multi_things_2.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            self.context.amp_init([mod0, mod1], [opt0, opt1]
+
+        # suppose the user wants one lr_scheduler auto-stepped every epoch...
+        self.lr_scheduler0 = context.lr_scheduler(
+            my_make_lr_scheduler(self.optimizer0), step_mode=STEP_EVERY_EPOCH
+        )
+
+        # and one scheduler stepped totally manually
+        self.lr_scheduler1 = context.lr_scheduler(
+            my_make_lr_scheduler(self.optimizer0), step_mode=MANUAL_STEP
```
Comment (rb-determined-ai) [inline pytorch_multi_things_2.py:194]: typo: this should use self.optimizer1

File: pytorch_multi_things_2.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        def my_clip_grads_1(...):
+            return ...
+
+        # This is analgous to the scaler.step() function
+        self.step(self.optimizer0, clip_grads=my_clip_grads_0)
+        self.step(self.optimizer1, clip_grads=my_clip_grads_1)
+
+        # The user opted to step lr_scheduler1 manually
+        if decide_to_step_lr_scheduler():
+            self.lr_scheduler1.step()
```
Comment (rb-determined-ai) [inline pytorch_multi_things_2.py:221]: I forgot to return a dictionary with metrics in the example

File: pytorch_multi_things_2.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        def my_clip_grads_0(...):
+            return ...
+
+        def my_clip_grads_1(...):
+            return ...
+
+        self.scaler.step(self.optimizer0, clip_grads=my_clip_grads_0)
+        self.scaler.step(self.optimizer1, clip_grads=my_clip_grads_1)
+
+        # We handle scaler.update()
```
Comment (rb-determined-ai) [inline pytorch_multi_things_2.py:162]: I forgot to return a dictionary of metrics in this example

File: pytorch_multi_things_2.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        with autocast():
+            output = self.model(input)
+            loss = loss_fn(output, target)
+
+        self.context.scale(loss).backward()
+
+        # Since we are going to support gradient aggregation internally, we might ignore this call.
+        self.context.step(self.optimizer)
+
+        # No need for the user to call scaler.update(); we'll handle that for them.
```
Comment (rb-determined-ai) [inline pytorch_multi_things_2.py:92]: I forgot to return a dictionary of metrics in this example

File: pytorch_multi_things_2.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+I found out that pytorch 1.6 will have its own amp module, and it's API is quite a bit cleaner than
+the apex.amp module (and it seems to use no gloabls :)
+
+It turns out that the pytorch-native amp API is pretty close to the API I proposed already, so I
+made some minor adjustemnts to make sure that they were even closer (and that we could easily
+support pytorch-native AMP in the future).
+
+The question I still have is: can we automatically handle optimizer.zero_grad()?  Is that too
+strong of an assumption?
```
Comment (rb-determined-ai) [inline pytorch_multi_things_2.py:19]: @aaron276h feels this is a reasonable assumption

File: pytorch_multi_things_2.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+        mod0 = self.context.model(my_make_model())
+        mod1 = self.context.model(my_make_model())
+        opt0 = self.context.optimizer(my_make_optimizer(mod0.parameters()))
+        opt1 = self.context.optimizer(my_make_optimizer(mod1.parameters()))
+
+        # This should work fine with apex.amp, and has the added benefit that you no longer have
+        # to configure amp via the experiment config... which in turn means you could do a
+        # hyperparameter search over it, if you wanted to.
+        (self.model0, self.model1), (self.optimizer0, self.optimizer1) = \
+            self.context.amp_init([mod0, mod1], [opt0, opt1]
```
Comment (rb-determined-ai) [inline pytorch_multi_things_2.py:185]: @aaron276h: we should expose many/all of the features of `amp.initialize()` in this call and get rid of the configuration entry for enabling mixed precision; this was one of the shortcomings of our current implementation.

File: pytorch_multi_things_2.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+        output1 = self.model1(input)
+        loss0 = loss_fn(2 * output0 + 3 * output1, target)
+        loss1 = loss_fn(3 * output0 - 5 * output1, target)
+
+        # TODO: name this better, but it is analagous to the scaler.scale() function
+        self.context.loss_tensor(loss0).backward(retain_graph=True)
+        self.context.loss_tensor(loss1).backward(retain_graph=True)
+
+        def my_clip_grads_0(...):
+            return ...
```
Comment (rb-determined-ai) [inline pytorch_multi_things_2.py:210]: @rb-determined-ai I don't like this strategy very much, but I think it is the best way we can do gradient clipping (if we are also doing gradient aggregation.

@aaron276h I don't like this, and I think calling a single gradient callback with a name argument corresponding to each optimizer would be better.

File: pytorch_multi_things_2.py [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+    def train_batch(self, batch):
+        input, target = batch
+
+        output0 = self.model0(input)
+        output1 = self.model1(input)
+        loss0 = loss_fn(2 * output0 + 3 * output1, target)
+        loss1 = loss_fn(3 * output0 - 5 * output1, target)
+
+        # TODO: name this better, but it is analagous to the scaler.scale() function
+        self.context.loss_tensor(loss0).backward(retain_graph=True)
```
Comment (rb-determined-ai) [inline pytorch_multi_things_2.py:206]: If we take this approach to mimic the scaler.scale() strategy we would have to add a backward hook to the tensor.  We would need to determine if that would interfere with the horovod backward hook, and make sure the order was always correct.

We all agreed that `loss_tensor()` was not an acceptable name for the final API, and that we could either rename it or go with the original `context.backward()` call where we call backward() ourselves.

We also agreed that in either case we are ok to move forward with customer discussions and address this detail at a later time.

File: pytorch_multi_things_2.py (hunk 1 above)
Comment (rb-determined-ai) [reply pytorch_multi_things_2.py:160]: > I think we could handle the optim.zero_grad ourselves especially since we handle gradient aggregation

Cool!  That is also what @aaron276h said, so if you both feel that way I feel extra good about this.

File: pytorch_multi_things_2.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+        input, target = batch
+
+        # I propose we handle zero_grad() automatically... is there a downside to this?
+        # self.optimizer.zero_grad()
+
+        with autocast():
+            output = self.model(input)
+            loss = loss_fn(output, target)
+
+        self.context.scale(loss).backward()
```
Comment (mackrorysd) [inline pytorch_multi_things_2.py:87]: If we're gonna handle scaler.step() and scaler.update() for them, why not also handle the scale(loss) call inside backward()

File: pytorch_multi_things_2.py [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+
+# The same thing in Determined:
+
+class MyPytorchTrial(PyTorchTrial):
+    def __init__(self, context):
+        # There's one context, just like there was one GradScaler()
+        # our context object will mimic the API of the GradScaler.
+        self.context = context
+
+        self.scaler = self.context.set_scaler(GradScaler())
```
Comment (mackrorysd) [inline pytorch_multi_things_2.py:71]: So it seems set_scaler() takes the place of amp_init()? There's nothing else that needs to be done (other than some safety checks similar to what we do on configuring Apex today, and setting _use_amp, etc. - and that can all just be done in set_scaler().

Comment (rb-determined-ai) [reply pytorch_multi_things_2.py:71]: I think the pytorch API is quite a bit more flexible than Apex (right?) in which case I'd rather we stick to the pytorch API, personally.  @katport has dealt with converting a pytorch amp model to apex amp, perhaps she has thoughts?

File: pytorch_multi_things_2.py (hunk 11 above)
Comment (rb-determined-ai) [reply pytorch_multi_things_2.py:87]: Shoot, it's been so long I can't remember the pytorch api I was imitating very well.  Off the top of my head, I would say that we should actually *avoid* doing things for the user in the domain of "data science" if the user can do it for themselves.  Whatever Horovod requires of us is sort of the limit to how far we can avoid doing things for the user, since "dtrain" falls squarely on us.

My guess in this case is that the user may have multiple losses and may want to scale some losses but not others.  Admittedly, I never fully understood what @autocast did so I'm not totally sure if it's unambiguous which losses need scaling and which don't.

Also maybe @liamcli or @katport might have some ideas here.
